routing/index_graph.cpp

```diff
@@ -11,22 +12,502 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+    return;
+  }
+
+  int shift = to.GetPointId() > from.GetPointId() ? 1 : -1;
+  for (int i = from.GetPointId(); i != to.GetPointId(); i += shift)
+    singleFeaturePath.emplace_back(from.GetFeatureId(), i);
+  singleFeaturePath.emplace_back(from.GetFeatureId(), to.GetPointId());
+}
+
+void IndexGraph::GetConnectionPaths(Joint::Id from, Joint::Id to,
+                                    vector<vector<RoadPoint>> & connectionPaths)
+{
+  CHECK_NOT_EQUAL(from, Joint::kInvalidId, ());
+  CHECK_NOT_EQUAL(to, Joint::kInvalidId, ());
+
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  CHECK_LESS(0, connections.size(), ());
+
+  connectionPaths.clear();
+  for (pair<RoadPoint, RoadPoint> const & c : connections)
+  {
+    vector<RoadPoint> roadPoints;
+    GetSingleFeaturePaths(c.first /* from */, c.second /* to */, roadPoints);
+    connectionPaths.push_back(move(roadPoints));
+  }
+}
+
+void IndexGraph::GetShortestConnectionPath(Joint::Id from, Joint::Id to,
+                                           vector<RoadPoint> & shortestConnectionPath)
+{
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  CHECK_LESS(0, connections.size(), ());
+
+  // Note. The case when size of connections is zero is the most common case. If not,
+  // it's necessary to perform a costy calls below.
+  if (connections.size() == 1)
+  {
+    GetSingleFeaturePaths(connections[0].first /* from */, connections[0].second /* to */,
+                          shortestConnectionPath);
+    return;
+  }
+
+  double minWeight = numeric_limits<double>::max();
+  pair<RoadPoint, RoadPoint> minWeightConnection;
+  for (pair<RoadPoint, RoadPoint> const & c : connections)
+  {
+    CHECK_EQUAL(c.first.GetFeatureId(), c.second.GetFeatureId(), ());
+    RoadGeometry const & geom = GetRoad(c.first.GetFeatureId());
+    if (!geom.IsRoad())
+      continue;
+
+    double weight = m_estimator->CalcEdgesWeight(geom, c.first.GetPointId() /* from */,
+                                                 c.second.GetPointId() /* to */);
+    if (weight < minWeight)
+    {
+      minWeight = weight;
+      minWeightConnection = c;
+    }
+  }
+
+  if (minWeight == numeric_limits<double>::max())
+  {
+    MYTHROW(RootException, ("Joint from:", from, "and joint to:", to,
+                            "are not connected a feature necessary type."));
+  }
+
+  GetSingleFeaturePaths(minWeightConnection.first /* from */, minWeightConnection.second /* to */,
+                        shortestConnectionPath);
+}
+
+void IndexGraph::GetFeatureConnectionPath(Joint::Id from, Joint::Id to, uint32_t featureId,
+                                          vector<RoadPoint> & featureConnectionPath)
+{
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  CHECK_LESS(0, connections.size(), ());
```

Comment:
```c++
CHECK(!connections.empty(), ());
```
BTW, what's wrong if connections is empty? This is possible, right? If so, why not to return false?